app/controllers/auth/actions/PayeAuthAction.scala

```diff
@@ -0,0 +1,112 @@
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.auth.DefaultAuthConnector
+import uk.gov.hmrc.play.http.HeaderCarrierConverter
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+
+class PayeAuthActionImpl @Inject() (
+  override val authConnector: DefaultAuthConnector,
+  cc: MessagesControllerComponents,
+  pertaxAuthService: PertaxAuthService
+)(implicit
+  ec: ExecutionContext,
+  appConfig: ApplicationConfig
+) extends PayeAuthAction
+    with AuthorisedFunctions
+    with Logging {
+
+  override val parser: BodyParser[AnyContent]               = cc.parsers.defaultBodyParser
+  override protected val executionContext: ExecutionContext = cc.executionContext
+
+  private val payeShuttered: Boolean = appConfig.payeShuttered
+
+  private def isAgent(enrolments: Set[Enrolment]) =
+    enrolments
+      .find(_.key == "IR-SA-AGENT")
+      .flatMap { enrolment =>
+        enrolment.identifiers
+          .find(id => id.key == "IRAgentReference")
+          .map(key => Uar(key.value))
+      }
+      .isDefined
+
+  override def invokeBlock[A](
+    request: Request[A],
+    block: PayeAuthenticatedRequest[A] => Future[Result]
+  ): Future[Result] =
+    if (payeShuttered) {
+      Future.successful(Redirect(controllers.paye.routes.PayeErrorController.serviceUnavailable))
+    } else {
+      implicit val hc: HeaderCarrier =
+        HeaderCarrierConverter.fromRequestAndSession(request, request.session)
+      pertaxAuthService.authorise[A, Request[A]](request).flatMap {
+        case None    =>
+          authorised(
+            ConfidenceLevel.L200 and AuthNino(hasNino = true) and CredentialStrength(CredentialStrength.strong)
+          ).retrieve(Retrievals.allEnrolments and Retrievals.nino and Retrievals.credentials) {
+            case Enrolments(enrolments) ~ Some(nino) ~ Some(credentials) =>
+              if (isAgent(enrolments)) {
+                Future.successful(Redirect(controllers.paye.routes.PayeErrorController.notAuthorised))
+              } else {
+                block {
+                  requests.PayeAuthenticatedRequest(
+                    nino = Nino(nino),
+                    credentials = credentials,
+                    request = request
+                  )
+                }
+              }
+            case _                                                       => throw new RuntimeException("Auth retrieval failed for user")
+          } recover {
+            case _: NoActiveSession =>
+              Redirect(
+                appConfig.payeLoginUrl,
+                Map(
+                  "continue_url" -> Seq(appConfig.payeLoginCallbackUrl),
+                  "origin"       -> Seq(appConfig.appName)
+                )
+              )
+            case NonFatal(e)        =>
+              logger.error(s"Exception in PayeAuthAction: $e", e)
+              Redirect(controllers.paye.routes.PayeErrorController.notAuthorised)
```

Comment:
No need. The error handler will catch that and use a service unavailable page

The session should be active. the pertax backend just checked that